File: android/src/main/java/com/ryltsov/alex/plugins/exif/ExifPlugin.java
package com.ryltsov.alex.plugins.exif;

import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Exif")
public class ExifPlugin extends Plugin {

    private static final String TAG = "ExifPlugin";

    private final Exif implementation = new Exif();

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void setCoordinates(final PluginCall call) {
        
        if (!call.getData().has("pathToImage")) {
            call.reject("Must provide an pathToImage");
            return;
        }
        if (!call.getData().has("lat")) {
            call.reject("Must provide an lat");
            return;
        }
        if (!call.getData().has("lng")) {
            call.reject("Must provide an lng");
            return;
        }

        String pathToImage = call.getString("pathToImage");
        double latitude = call.getDouble("lat");
        double longitude = call.getDouble("lng");

        try {
            implementation.setCoordinates(pathToImage, latitude, longitude);
        } catch (Exception e) {
            Log.e(TAG, "Error setting GPS data to image", e);
            call.reject(e.getLocalizedMessage(), null, e);
            return;
        }

        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void getCoordinates(final PluginCall call) {
        
        if (!call.getData().has("pathToImage")) {
            call.reject("Must provide an pathToImage");
            return;
        }

        String pathToImage = call.getString("pathToImage");

        double[] latLong = null;
        try {
            latLong = implementation.getCoordinates(pathToImage);
        } catch (Exception e) {
            Log.e(TAG, "Error getting GPS data from image", e);
            call.reject(e.getLocalizedMessage(), null, e);
            return;
        }
        if ((latLong != null ? latLong.length : 0) == 2) {
            JSObject ret = new JSObject();
            double latitude = latLong[0];
            double longitude = latLong[1];
            ret.put("lat", latitude);
            ret.put("lng", longitude);
            call.resolve(ret);
            return;
        }
        call.resolve();
    }

}

File: ios/Sources/ExifPlugin/ExifPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ExifPlugin)
public class ExifPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "ExifPlugin"
    public let jsName = "Exif"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setCoordinates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCoordinates", returnType: CAPPluginReturnPromise)
    ]

    // Message constants
    static let invalidUrlError = "Invalid URL"
    static let failedToLoadImageError = "Failed to load image"
    static let failedToCreateDestinationImageError = "Failed to create destination image"
    static let failedToSaveImageError = "Failed to save image"

    private let implementation = Exif()

    @objc func setCoordinates(_ call: CAPPluginCall) {

        guard let pathToImage = call.options["pathToImage"] as? String else {
            call.reject("Must provide an pathToImage")
            return
        }
        guard let latitude = call.options["lat"] as? Double else {
            call.reject("Must provide an lat")
            return
        }
        guard let longitude = call.options["lng"] as? Double else {
            call.reject("Must provide an lng")
            return
        }

        do {
            try implementation.setCoordinates(pathToImage, latitude, longitude)
            call.resolve()
        } catch ImageProcessingError.invalidURL {
            call.reject(ExifPlugin.invalidUrlError)
        } catch ImageProcessingError.failedToLoadImage {
            call.reject(ExifPlugin.failedToLoadImageError)
        } catch ImageProcessingError.failedToCreateDestinationImage {
            call.reject(ExifPlugin.failedToCreateDestinationImageError)
        } catch ImageProcessingError.failedToSaveImage {
            call.reject(ExifPlugin.failedToSaveImageError)
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }

    }

    @objc func getCoordinates(_ call: CAPPluginCall) {

        guard let pathToImage = call.options["pathToImage"] as? String else {
            call.reject("Must provide an pathToImage")
            return
        }

        do {
            let coordinates = try implementation.getCoordinates(filePath: pathToImage)
            call.resolve([
                "lat": coordinates.latitude,
                "lng": coordinates.longitude
            ])
        } catch ImageProcessingError.invalidURL {
            call.reject(ExifPlugin.invalidUrlError)
        } catch ImageProcessingError.failedToLoadImage {
            call.reject(ExifPlugin.failedToLoadImageError)
        } catch ImageProcessingError.noGPSData {
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }

    }

}
